### DISCUSSION TITLE
Unexpected bond dimension

### DISCUSSION THREAD

**Discussion by Shovan-Physics** [block-hczhai/block2-preview#126] [Q&A] [answered]
I find that `mps.info.get_max_bond_dimension()` is sometimes larger than the maximum bond dimension allowed during sweeps. 

I am looking at the SU(2) t-J model on a ladder, using the example in https://block2.readthedocs.io/en/latest/tutorial/custom-hamiltonians.html#SU(2)-t-J-Model and the following parameters:

```
LX, LY = 20, 2
L = LX * LY
J = 0.0
N_ELEC = L-1
TWO_S = L-1

bond_dims = [100] * 4 + [200] * 4
```
I get a ground-state energy `E = -2.97766` and `mps.info.get_max_bond_dimension() = 441`, even though the maximum allowed bond dimension was 200.

If I use the modified symmetry type suggested here: https://github.com/block-hczhai/block2-preview/discussions/115#discussioncomment-10258650 (required for converting to SZ), I get an even higher bond dimension for the same parameters, with `mps.info.get_max_bond_dimension() = 860`.

The mismatch persists for nonzero `J` and lower values of `TWO_S`.

How do I understand this? Is there a way to track the actual (maximum) bond dimension at the end of each sweep?

**Comment by hczhai** [comment, marked as answer]
Thanks for reporting this problem. The value returned from ``mps.info.get_max_bond_dimension()`` is incorrect.

As a workaround, you can add an argument ``full_fci=False`` to ``driver.get_random_mps``, then ``mps.info.get_max_bond_dimension()`` will behave correctly.

This will be fixed in version ``0.5.3rc19`` so that ``mps.info.get_max_bond_dimension()`` will print the correct value with or without ``full_fci=False``.

**Reply by Shovan-Physics** [reply]
Thanks for the quick response. 

As a related question, does `block2` have any ''observer" feature that can be used to track the actual bond dimension or some expectation value during the sweeps?

**Reply by hczhai** [reply]
After you upgrade ``block2`` to version ``0.5.3rc19`` (with ``pip`` options ``--upgrade --force-reinstall --no-deps --no-cache-dir``), you can add the following before ``driver.dmrg``:

```python3
mps = ...
mpo = ...

@driver.set_callback
def observer(event, iprint):
    if event == "DMRG::sweep.end" and iprint > 0:
        print('Sweep Mmps =', mps.info.get_max_bond_dimension(), 'Expt =', driver.expectation(mps, mpo, mps))

driver.dmrg(...)
```

Then at the end of each sweep, the ``observer`` function will be executed.

**Reply by Shovan-Physics** [reply]
Excellent, thanks.
